### ISSUE TITLE
Issue when installing node modules

### ISSUE THREAD

**Issue by tiagojsag** [hotosm/osm-analytics-cruncher#12]
Hi everyone,

I'm trying to set up the project on my local env, and I'm getting the message bellow. This is taken from a macOS, but I am actually getting the same issue on arch linux. It seems like one of the dependencies is misbehaving, but I am not sure which. I was hoping someone around here might be able to point me in the right direction.

Thanks 

```
CXX(target) Release/obj.target/gdal/src/utils/fast_buffer.o
In file included from ../src/utils/fast_buffer.cpp:4:
In file included from ../src/utils/fast_buffer.hpp:11:
In file included from ../node_modules/nan/nan.h:194:
../node_modules/nan/nan_maybe_43_inl.h:221:17: warning: 'CloneElementAt' is deprecated [-Wdeprecated-declarations]
  return array->CloneElementAt(GetCurrentContext(), index);
                ^
/Users/tiagojsag/.node-gyp/7.10.0/include/node/v8.h:3347:36: note: 'CloneElementAt' has been explicitly marked deprecated here
                MaybeLocal<Object> CloneElementAt(Local<Context> context,
                                   ^
In file included from ../src/utils/fast_buffer.cpp:4:
In file included from ../src/utils/fast_buffer.hpp:11:
In file included from ../node_modules/nan/nan.h:200:
In file included from ../node_modules/nan/nan_new.h:189:
../node_modules/nan/nan_implementation_12_inl.h:40:29: warning: 'New' is deprecated [-Wdeprecated-declarations]
  return v8::BooleanObject::New(value).As<v8::BooleanObject>();
                            ^
/Users/tiagojsag/.node-gyp/7.10.0/include/node/v8.h:4490:56: note: 'New' has been explicitly marked deprecated here
  V8_DEPRECATED("Pass an isolate", static Local<Value> New(bool value));
                                                       ^
In file included from ../src/utils/fast_buffer.cpp:4:
In file included from ../src/utils/fast_buffer.hpp:11:
../node_modules/nan/nan.h:602:20: error: no type named 'GCEpilogueCallback' in 'v8::Isolate'
      v8::Isolate::GCEpilogueCallback callback
      ~~~~~~~~~~~~~^
../node_modules/nan/nan.h:608:20: error: no type named 'GCEpilogueCallback' in 'v8::Isolate'
      v8::Isolate::GCEpilogueCallback callback) {
      ~~~~~~~~~~~~~^
../node_modules/nan/nan.h:613:20: error: no type named 'GCPrologueCallback' in 'v8::Isolate'
      v8::Isolate::GCPrologueCallback callback
      ~~~~~~~~~~~~~^
../node_modules/nan/nan.h:619:20: error: no type named 'GCPrologueCallback' in 'v8::Isolate'
      v8::Isolate::GCPrologueCallback callback) {
      ~~~~~~~~~~~~~^
../node_modules/nan/nan.h:1955:15: warning: 'SetAccessor' is deprecated [-Wdeprecated-declarations]
  return obj->SetAccessor(
              ^
/Users/tiagojsag/.node-gyp/7.10.0/include/node/v8.h:3027:22: note: 'SetAccessor' has been explicitly marked deprecated here
                bool SetAccessor(Local<Name> name,
                     ^
../src/utils/fast_buffer.cpp:18:50: warning: 'NewInstance' is deprecated [-Wdeprecated-declarations]
        Local<Object> actualBuffer = bufferConstructor->NewInstance(3, constructorArgs);
                                                        ^
/Users/tiagojsag/.node-gyp/7.10.0/include/node/v8.h:3658:31: note: 'NewInstance' has been explicitly marked deprecated here
                Local<Object> NewInstance(int argc, Local<Value> argv[]) const);
                              ^
4 warnings and 4 errors generated.
make: *** [Release/obj.target/gdal/src/utils/fast_buffer.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/tiagojsag/.nvm/versions/node/v7.10.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:194:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Darwin 16.5.0
gyp ERR! command "/Users/tiagojsag/.nvm/versions/node/v7.10.0/bin/node" "/Users/tiagojsag/.nvm/versions/node/v7.10.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build" "--fallback-to-build" "--module=/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/lib/binding/gdal.node" "--module_name=gdal" "--module_path=/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/lib/binding"
gyp ERR! cwd /Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal
gyp ERR! node -v v7.10.0
gyp ERR! node-gyp -v v3.5.0
gyp ERR! not ok
node-pre-gyp ERR! build error
node-pre-gyp ERR! stack Error: Failed to execute '/Users/tiagojsag/.nvm/versions/node/v7.10.0/bin/node /Users/tiagojsag/.nvm/versions/node/v7.10.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/lib/binding/gdal.node --module_name=gdal --module_path=/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/lib/binding' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
node-pre-gyp ERR! stack     at emitTwo (events.js:106:13)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:194:7)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:899:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
node-pre-gyp ERR! System Darwin 16.5.0
node-pre-gyp ERR! command "/Users/tiagojsag/.nvm/versions/node/v7.10.0/bin/node" "/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal
node-pre-gyp ERR! node -v v7.10.0
node-pre-gyp ERR! node-pre-gyp -v v0.6.15
node-pre-gyp ERR! not ok
Failed to execute '/Users/tiagojsag/.nvm/versions/node/v7.10.0/bin/node /Users/tiagojsag/.nvm/versions/node/v7.10.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/lib/binding/gdal.node --module_name=gdal --module_path=/Users/tiagojsag/web/osm-analytics-cruncher/node_modules/gdal/lib/binding' (1)
osm-analytics-cruncher@0.0.0 /Users/tiagojsag/web/osm-analytics-cruncher
└─┬ turf@2.0.2
  └─┬ turf-triangle-grid@1.0.1
    └─┬ tape@3.6.1
      ├── deep-equal@0.2.2
      ├── defined@0.0.0
      ├─┬ glob@3.2.11
      │ └─┬ minimatch@0.3.0
      │   ├── lru-cache@2.7.3
      │   └── sigmund@1.0.1
      ├── object-inspect@0.4.0
      └── resumer@0.0.0

npm ERR! Darwin 16.5.0
npm ERR! argv "/Users/tiagojsag/.nvm/versions/node/v7.10.0/bin/node" "/Users/tiagojsag/.nvm/versions/node/v7.10.0/bin/npm" "i"
npm ERR! node v7.10.0
npm ERR! npm  v4.2.0
npm ERR! code ELIFECYCLE
npm ERR! errno 1

npm ERR! gdal@0.8.0 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the gdal@0.8.0 install script 'node-pre-gyp install --fallback-to-build'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the gdal package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs gdal
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls gdal
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/tiagojsag/.npm/_logs/2017-05-14T10_38_01_376Z-debug.log
```

**Comment by cgiovando** [comment]
@tyrasd @davidneudorfer have you had any similar problem when installing locally?

**Comment by tyrasd** [comment]
I think I had some similar issues once. It's likely an upstream issue (see https://github.com/mapbox/mapbox-tile-copy/issues/47) that is eventually related to the relatively large node-mapnik library which is apparently not compatible with very many node versions. :man_shrugging: 

Can you try commenting out the *"mapbox-tile-copy"* dependency [here](https://github.com/hotosm/osm-analytics-cruncher/blob/master/package.json#L28)? That dependency is only needed in [hotprojects.sh](https://github.com/hotosm/osm-analytics-cruncher/blob/master/hotprojects.sh), which is only a secondary output of the cruncher and independent from the main functionality of generating the osm-ananlytics vector tiles.

PS: If you're starting to work on this, it would probably make sense to directly start with the [notippecanoe](https://github.com/hotosm/osm-analytics-cruncher/tree/notippecanoe) branch, which features a pure nodejs-implementation of the functionality one had to manually build and install *tippecanoe* for. It's also quite a bit faster. :)

**Comment by nerik** [comment]
Thanks for the pointers. Here's how I got to test the cruncher on my machine:

1. switch to node v4.6.0 (use nvm)
```
nvm install v4.6.0
nvm use v4.6.0
```
(note: the `nvm use` dance could later be avoided by creating a .nvmrc file containing `v4.6.0`, and set up your shell to switch node versions automatically when cding: https://github.com/creationix/nvm#calling-nvm-use-automatically-in-a-directory-with-a-nvmrc-file)

2. switch to `notippecanoe` branch
3. `npm i`
4. `npm i -S tile-reduce`
5. `chmod +x run.sh`
6. in run.sh, set up RESULTS_DIR to an existing folder
7. instead of letting the script downloading the whole planet file, it is recommended to get a country subset from https://osmlab.github.io/osm-qa-tiles/country.html
8. put the downloaded file at the root of the project as `planet.mbtiles`
9. Comment out a few steps from run.sh, from line 20 to 36 (skips npm install,  cleanup, hotprojects.sh, and downloading the planet file)
10. Comment out lines 45, 52, 59. They remove files that seemingly are never created.
11. Comment out  lines 47, 54, 61. They use forever which failed for me, I have yet to investigate
12. `./run.sh`
13. You should get mbtiles files, ready to be served for the client.

Later on I will make a few PR's to address some of the issues I found in the whole process (I also think the tiles generation part should be extracted from run.sh), but feel free to comment on the steps above if something seems amiss.

**Comment by tyrasd** [comment]
:+1:

> switch to node v4.6.0

Any particular reason for this? If I remember correctly, the notippecanoe branch should work on newer node versions as well.

> Comment out lines 45, 52, 59. They remove files that seemingly are never created.

It's supposed to remove existing files in the result directory (e.g. when running the cruncher daily with fresh data). I guess this could also be achieved by using `mv -f` in the respective lines below.

> Comment out lines 47, 54, 61. They use forever …

Yeah, it's assumed that there's a *forever* script already running and [forever](https://www.npmjs.com/package/forever) to be installed globally (see on the top: *SERVER_SCRIPT - node script that serves the .mbtiles. Assumed to be already started with forever*). I think this could be improved by putting the full command into the bash variable, so that arbitrary commands can be executed after each output .mbtiles file has been generated.

**Comment by nerik** [comment]
Thanks. Here's a PR addressing some of these issues https://github.com/GFDRR/osm-analytics-cruncher/pull/3

- node version: yes you're right actually, tested and working with node v7.5.0.
- fixed removing files issue with the PR

I'm still unclear why the final `./[feature_type].mbtiles` files get deleted. Are you using tilelive on production, or generated PBF tiles ?

**Comment by tyrasd** [comment]
> Are you using tilelive on production, or generated PBF tiles?

We're currently using a tilelive based [script](https://github.com/hotosm/osm-analytics-cruncher/blob/master/server/serve.js).

> I'm still unclear why the final ./[feature_type].mbtiles files get deleted

Well, first a new .mbtiles files is produced for the current planet, which then gets copied over to the results directory. After that the old .mbtiles file gets deleted and the new one renamed to the final filename. That way there is almost no delay between dropping the old data and putting the updated data in.

**Comment by tyrasd** [comment]
this should be fixed with 8375a66032d72e1791735cb985863e77095bda7f
